### ISSUE TITLE
False positive - `InvalidParamDefault` vs union types vs ArrayObject

### ISSUE THREAD

**Issue by mkilmanas** [vimeo/psalm#6668]
Psalm version: `Psalm 4.10.0@916b098b008f6de4543892b1e0651c1c3b92cbfa` & `Psalm dev-master@6abdb09`
PHP version:  `PHP 8.0.11 (cli) (built: Sep 23 2021 21:26:06) ( NTS )`

Returned error:
```
ERROR: InvalidParamDefault - src/Dto/Output/ResponseItem.php:18:9 - Default value type array<empty, empty> for argument 9 of method App\Dto\Output\ResponseItem::__construct does not match the given type ArrayObject<array-key, mixed> (see https://psalm.dev/062)
        public \ArrayObject | array $meta = [],
```

Relevant code:
```php
<?php

declare(strict_types=1);

namespace App\Dto\Output;

class ResponseItem
{
    public function __construct(
        public string $calculatedDeliveryPredictionItemId,
        public string $productId,
        public \ArrayObject | array $options,
        public int $quantity,
        public \DateTimeInterface $minDate,
        public \DateTimeInterface $maxDate,
        public string $status,
        public array $tags = [],
        public \ArrayObject | array $meta = [],
    ) {
    }
}
```

It appears that Psalm tries to validate the default value of `$meta = []` against ArrayObject ignoring all the other types of the union. E.g.
- `array | \ArrayObject $meta = []` -> also fails
- `string | int $meta = 5` -> works OK
- `string | array $meta = []` -> works OK
- '\DateTimeInterface | array $meta = []` -> works OK
- `\ArrayObject | array | null $meta = null` -> works OK (!!)

So it seems that `ArrayObject` vs `array` is a special case.

Strangely enough it does not seem to complain about it on https://psalm.dev/r/706023893a even though it's the same `6abdb09` version.

my `psalm.xml` (baseline is in fact empty):
```xml
<?xml version="1.0"?>
<psalm
    errorLevel="4"
    phpVersion="8.0"
    cacheDirectory="var/cache/psalm"
    resolveFromConfigFile="true"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xmlns="https://getpsalm.org/schema/config"
    xsi:schemaLocation="https://getpsalm.org/schema/config vendor/vimeo/psalm/config.xsd"
    allowPhpStormGenerics="true"
    errorBaseline="psalm-baseline.xml"
>
    <projectFiles>
        <directory name="src/" />
        <ignoreFiles>
            <directory name="vendor" />
        </ignoreFiles>
    </projectFiles>

    <fileExtensions>
        <extension name=".php" />
    </fileExtensions>
</psalm>
```

**Comment by orklah** [comment]
allowPhpStormGenerics is messing with the type. When we enable it on psalm.dev, we can reproduce: https://psalm.dev/r/cbde4d0046

However, it was designed for phpdoc at the time. We didn't have unions in signature back then. It should not collapse the type in signature

**Comment by orklah** [comment]
Good bug report BTW :)

**Comment by mkilmanas** [comment]
Wow that was quick turnaround - thanks a bunch!
